contracts/exchange-forwarder/contracts/src/libs/LibAssetDataTransfer.sol

```diff
@@ -0,0 +1,258 @@
+    /// @param from Address to transfer asset from.
+    /// @param to Address to transfer asset to.
+    /// @param amount Amount of asset to transfer to sender.
+    function transferFrom(
+        bytes memory assetData,
+        address from,
+        address to,
+        uint256 amount
+    )
+        internal
+    {
+        if (amount == 0) {
+            return;
+        }
+
+        bytes4 proxyId = assetData.readBytes4(0);
+
+        if (
+            proxyId == IAssetData(address(0)).ERC20Token.selector ||
+            proxyId == IAssetData(address(0)).ERC20Bridge.selector
+        ) {
+            assetData._transferERC20Token(
+                from,
+                to,
+                amount
+            );
+        } else if (proxyId == IAssetData(address(0)).ERC721Token.selector) {
+            assetData._transferERC721Token(
+                from,
+                to,
+                amount
+            );
+        } else if (proxyId == IAssetData(address(0)).ERC1155Assets.selector) {
+            assetData._transferERC1155Assets(
+                from,
+                to,
+                amount
+            );
+        } else if (proxyId == IAssetData(address(0)).MultiAsset.selector) {
+            assetData._transferMultiAsset(
+                from,
+                to,
+                amount
+            );
+        } else if (proxyId != IAssetData(address(0)).StaticCall.selector) {
+            LibRichErrors.rrevert(LibForwarderRichErrors.UnsupportedAssetProxyError(
+                proxyId
+            ));
+        }
+    }
+
+    ///@dev Transfer asset from sender to this contract.
+    /// @param assetData Byte array encoded for the respective asset proxy.
+    /// @param amount Amount of asset to transfer to sender.
+    function transferIn(
+        bytes memory assetData,
+        uint256 amount
+    )
+        internal
+    {
+        assetData.transferFrom(
+            msg.sender,
+            address(this),
+            amount
+        );
+    }
+
+    ///@dev Transfer asset from this contract to sender.
+    /// @param assetData Byte array encoded for the respective asset proxy.
+    /// @param amount Amount of asset to transfer to sender.
+    function transferOut(
+        bytes memory assetData,
+        uint256 amount
+    )
+        internal
+    {
+        assetData.transferFrom(
+            address(this),
+            msg.sender,
+            amount
+        );
+    }
+
+    /// @dev Decodes ERC20 or ERC20Bridge assetData and transfers given amount to sender.
+    /// @param assetData Byte array encoded for the respective asset proxy.
+    /// @param from Address to transfer asset from.
+    /// @param to Address to transfer asset to.
+    /// @param amount Amount of asset to transfer to sender.
+    function _transferERC20Token(
+        bytes memory assetData,
+        address from,
+        address to,
+        uint256 amount
+    )
+        internal
```

Comment:
Could these just as well be `private`?

Reply:
I'm fine with it either way 🤷‍♂ It's possible we want to directly access these if we're limiting support to certain asset types.

Reply:
Then should we drop the underscore?